module/VuFind/src/VuFind/Controller/CheckoutsController.php

```diff
@@ -0,0 +1,271 @@
+            if ($pageOptions['ilsPaging'] || ($i >= $pageStart && $i <= $pageEnd)) {
+                $driversNeeded[] = $current;
+            } else {
+                $hiddenTransactions[] = $current;
+            }
+            if ($purgeSelectedAllowed && isset($current['row_id'])) {
+                $this->rememberValidRowId($current['row_id']);
+            }
+        }
+
+        $transactions = $this->ilsRecords()->getDrivers($driversNeeded);
+        $sortList = $pageOptions['sortList'];
+        $params = $pageOptions['ilsParams'];
+        return $this->createViewModel(
+            compact(
+                'transactions',
+                'paginator',
+                'params',
+                'hiddenTransactions',
+                'sortList',
+                'functionConfig',
+                'purgeAllAllowed',
+                'purgeSelectedAllowed'
+            )
+        );
+    }
+
+    /**
+     * Purge loans from loan history
+     *
+     * @return mixed
+     */
+    public function purgeHistoryAction()
+    {
+        // Stop now if the user does not have valid catalog credentials available:
+        if (!is_array($patron = $this->catalogLogin())) {
+            return $patron;
+        }
+
+        $redirectResponse = $this->redirect()->toRoute('checkouts-listhistory');
+        if ($this->formWasSubmitted('purgeSelected', false)
+            || $this->formWasSubmitted('purgeAll', false)
+        ) {
+            $csrfToken = $this->getRequest()->getPost()->get('csrf');
+            if (!$this->csrf->isValid($csrfToken)) {
+                $this->flashMessenger()
+                    ->addErrorMessage('error_inconsistent_parameters');
+                return $redirectResponse;
+            }
+            // After successful token verification, clear list to shrink session:
+            $this->csrf->trimTokenList(0);
+            $catalog = $this->getILS();
+            if ($this->formWasSubmitted('purgeAll', false)) {
+                $result = $catalog->purgeTransactionHistory($patron, null);
+            } else {
+                $ids = $this->getRequest()->getPost()->get('purgeSelectedIDs', []);
+                if (!$this->validateRowIds($ids)) {
+                    $this->flashMessenger()
+                        ->addErrorMessage('error_inconsistent_parameters');
+                    return $redirectResponse;
+                }
+                $result = $catalog->purgeTransactionHistory($patron, $ids);
+            }
+            $this->flashMessenger()->addMessage(
+                $result['status'],
+                $result['success'] ? 'info' : 'error'
+            );
+        }
+        return $redirectResponse;
+    }
+
+    /**
+     * Return a session container for validating selected row ids.
+     *
+     * @return \Laminas\Session\Container
+     */
+    protected function getRowIdContainer()
+    {
+        return new \Laminas\Session\Container('row_ids', $this->sessionManager);
```

Comment:
Would it be better to store this in a property and/or build this in the constructor? Otherwise we're building a new object on every call... which is functionally okay but seems inefficient.